gcm/sub.go

```diff
@@ -0,0 +1,347 @@
+package gcm
+
+import (
+	"encoding/json"
+	"errors"
+	log "github.com/Sirupsen/logrus"
+	"github.com/alexjlockwood/gcm"
+	"github.com/smancke/guble/protocol"
+	"github.com/smancke/guble/server"
+	"github.com/smancke/guble/store"
+	"math"
+	"strconv"
+	"strings"
+)
+
+const (
+	// default subscription channel buffer size
+	subBufferSize = 50
+)
+
+var (
+	errSubReplaced = errors.New("Subscription replaced")
+)
+
+type jsonError struct {
+	json string
+}
+
+func (e *jsonError) Error() string {
+	return e.json
+}
+
+// Creates a subscription and returns the pointer
+func newSub(gcm *Connector, route *server.Route, lastID uint64) *sub {
+	return &sub{
+		gcm:    gcm,
+		route:  route,
+		lastID: lastID,
+		logger: logger.WithFields(log.Fields{
+			"gcmID":  route.ApplicationID,
+			"userID": route.UserID,
+			"topic":  string(route.Path),
+		}),
+	}
+}
+
+// creates a subscription and adds it in router/kvstore then starts listening for messages
+func initSub(gcm *Connector, topic, userID, gcmID string, lastID uint64) (*sub, error) {
+	route := server.NewRoute(topic, gcmID, userID, subBufferSize)
+	s := newSub(gcm, route, 0)
+	if err := s.store(); err != nil {
+		return nil, err
+	}
+	s.start()
+	return s, nil
+}
+
+// sub represent a GCM subscription
+type sub struct {
```

Comment:
maybe rename it in subscription
and also initSub would be IMHO initSubcriptions